File: src/emitter/mermaid/render-highlight.test.ts
import { describe, expect, test } from "vitest";

import { renderHighlight } from "./render-highlight.js";
import { darkTheme } from "./theme/dark-theme.js";

describe("renderHighlight", () => {
  test("writes nothing when no ids are highlighted", () => {
    const lines: /* mutable */ string[] = [];
    renderHighlight(new Set<string>(), [], darkTheme, lines);
    expect(lines).toEqual([]);
  });

  test("emits one `style` line per highlighted node", () => {
    const lines: /* mutable */ string[] = [];
    renderHighlight(new Set(["n_a"]), [], darkTheme, lines);
    expect(lines).toEqual([
      `  style n_a fill:${darkTheme.highlight.fill},stroke:${darkTheme.highlight.stroke},color:${darkTheme.highlight.color};`,
    ]);
  });

  test("emits a single linkStyle line covering every highlighted edge index", () => {
    const lines: /* mutable */ string[] = [];
    renderHighlight(new Set(["n_a"]), [0, 2, 3], darkTheme, lines);
    expect(lines.at(-1)).toEqual(
      `  linkStyle 0,2,3 stroke:${darkTheme.highlight.edgeStroke},stroke-width:${darkTheme.highlight.edgeStrokeWidth};`,
    );
  });

  test("skips the linkStyle line when no edge indices are supplied", () => {
    const lines: /* mutable */ string[] = [];
    renderHighlight(new Set(["n_a"]), [], darkTheme, lines);
    expect(lines.some((v) => v.startsWith("  linkStyle"))).toEqual(false);
  });
});

File: crates/unsnarl-emitter-mermaid/src/render_highlight_test.rs
use super::render_highlight;
use crate::theme::DARK_THEME;

#[test]
fn writes_nothing_when_no_ids_are_highlighted() {
    let mut lines: Vec<String> = Vec::new();
    render_highlight(&[], &[], &DARK_THEME, &mut lines);
    assert!(lines.is_empty());
}

#[test]
fn emits_one_style_line_per_highlighted_node() {
    let ids = vec!["n_a".to_string()];
    let mut lines: Vec<String> = Vec::new();
    render_highlight(&ids, &[], &DARK_THEME, &mut lines);
    let h = &DARK_THEME.highlight;
    assert_eq!(
        lines,
        vec![format!(
            "  style n_a fill:{},stroke:{},color:{};",
            h.fill, h.stroke, h.color
        )]
    );
}

#[test]
fn emits_a_single_link_style_line_covering_every_highlighted_edge_index() {
    let ids = vec!["n_a".to_string()];
    let mut lines: Vec<String> = Vec::new();
    render_highlight(&ids, &[0, 2, 3], &DARK_THEME, &mut lines);
    let h = &DARK_THEME.highlight;
    let last = lines.last().expect("last line present");
    assert_eq!(
        last,
        &format!(
            "  linkStyle 0,2,3 stroke:{},stroke-width:{};",
            h.edge_stroke, h.edge_stroke_width
        )
    );
}

#[test]
fn skips_the_link_style_line_when_no_edge_indices_are_supplied() {
    let ids = vec!["n_a".to_string()];
    let mut lines: Vec<String> = Vec::new();
    render_highlight(&ids, &[], &DARK_THEME, &mut lines);
    assert!(!lines.iter().any(|v| v.starts_with("  linkStyle")));
}
